### ISSUE TITLE
Regent: index launch with interfering index spaces

### ISSUE THREAD

**Issue by CharlelieLrt** [StanfordLegion/legion#1312]
In Regent, I am trying to have an index launch executing in parallel a task taking as arguments both a region `f` and a sub-region `p[c]` of `f`, such as:

```
var is = ispace(int3d, {n, m, p})
var p_is = ispace(int3d, {N,1,1})
var f = region(is, my_field_space)
var p = partition(equal, f, p_is)
...
__demand(__index_launch)
for c in p_is do
   my_task(p[c], f)
end
```

where `my_task` has the following structure:

```
__demand(__inner)
task my_task(p_i: region(ispace(int3d), my_field_space), f: region(ispace(int3d), my_field_space))
where reads writes (p_i), reads(f) do
...
write_to(p_i)
...
read_from(f)
...
end
```

where `write_to` is a task that writes to `p[c]` and `read_from` is a task reading from the region `f`.
I was expecting the top-level loop to be executed in parallel over multiple GPUs, in particular the `write_to(p_i)` tasks. But it seems that the "inter-dependency" between the index spaces of `p[c]` and `f` prevents the tasks to run in parallel.
Using `legion_prof`, I am seeing GPU#0 with a very high utilization, and running all the `write_to()` tasks, while all the other GPUs remain idle.
Also, I did not notice any warning or error at compile or run time, but `legion_spy` returns the following error:

```
Region requirements 0 and 1 of operation my_task are interfering in main
ERROR: Found interfering index space launch: my_task!
```

This suggests that I am doing something wrong.
Would it be possible to have a confirmation that it is indeed forbidden to have index launches with interfering index spaces, such as in the example above? If, so any workaround would be very much appreciated.

**Comment by lightsighter** [comment]
Can you upload the Legion Spy log file here? Legion will also catch this if it is compiled in debug mode (a flag command flag will be available for detecting this even in release mode in the nearish future). Legion Spy in general is very good about recognizing these things.

As for whether you're doing something wrong or not, I think Regent is supposed to help you out by creating the region requirements in a way that is non-interfering, but we'll need to wait for @elliottslaughter to say for sure.

**Comment by CharlelieLrt** [comment]
Here are the Spy log files:
[spy_log.tar.gz](https://github.com/StanfordLegion/legion/files/9404547/spy_log.tar.gz)

**Comment by elliottslaughter** [comment]
To me it looks like we have a task with RW and RO privileges on two overlapping regions. (The fact that they happen to be parent and child isn't important here.)

What do you expect Regent to do here? As far as I understand, this is a violation of Legion semantics. Without doing partitioning (which in general, Regent is never going to do on behalf of the user), there is nothing Regent can do to make this situation legal. The best it can do is make sure the index launch optimization is disabled---which it looks like Regent is indeed doing---but when even a single task launch is illegal there's not much we can do.

@CharlelieLrt, in order to make this situation work you'll either need to:

 1. Create multiple fields so that the region requirements for the parent and child can be non-interfering (i.e., one reads field A while another writes field B).
 2. Partition the parent so that you have the subset R and a region NOT_R which is the parent minus R. *But then you will not be able to index launch this task.*
 3. Talk to us further about what you're trying to accomplish so we can help you reach a design which is actually parallel. Because the code you've got here is not, no matter what kind of magic Legion attempts to do.

**Comment by unpyoukz** [comment]
Just to chime in, from an application perspective to handle shared data among my_task (where the question of @CharlelieLrt came out I believe), there is a separate effort including a mapper (we will perhaps reach out in an existing/separate post). Nevertheless I also believe this discussion would be valuable especially if there is a way to automate the parallelization when both W and R are called over f multiple times in leaf tasks of my_task but R is much less frequent than W.

**Comment by CharlelieLrt** [comment]
As @unpyoukz mentioned there might be a workaround for our specific application, using a custom mapper.
Also my initial question was probably too generic, so I am closing.

**Comment by elliottslaughter** [comment]
Just to clarify one point: the code posted at the top of this issue cannot be made to work with any mapper.

So perhaps what you mean to say is that your original problem (from which the code at top was reduced from) is such that a mapper customization could plausibly fix it. Because otherwise I'm afraid you may be working from a flawed understanding of what the problem is.

**Comment by CharlelieLrt** [comment]
I fully agree with that.
The code snippet in the initial post was over-simplified in comparison to what is used in our application.
I'll try the mapper proposed by @unpyoukz and if the issue persists, I'll follow up with a minimum working example more representative of our use case.
